### PR TITLE
Symbolic speedup

### PR DESCRIPTION
**Pull request by mloubout**
Minor tweaks to improve the speed of `evaluate` that is quite slow going through all the nobs of sympy. Bypasses a lot of it via direct call to `_subs` and avoid very expensive `evalf` on expression while only needed on the coefficients.

timing improvement for 12th order 3D on my laptop for this branch

```
In [1]: %timeit u.laplace.laplace.evaluate                                                                                                                                           
1.04 s ± 71.5 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

In [2]: %timeit u.laplace.laplace.evaluate                                                                                                                                           
1.06 s ± 132 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
```
vs master:

```
In [1]: %timeit u.laplace.laplace.evaluate   
6.23 s ± 1.02 s per loop (mean ± std. dev. of 7 runs, 1 loop each)

In [2]: %timeit u.laplace.laplace.evaluate                                                                                                                                           
7.16 s ± 2.23 s per loop (mean ± std. dev. of 7 runs, 1 loop each)

In [3]: %timeit u.laplace.laplace.evaluate                                                                                                                                           
5.56 s ± 268 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

```

And aside, `setup.cfg` updated for `flake8-nb` if someone wants to play with it but not adding it to CI.

### CODE REVIEW

File: devito/finite_differences/differentiable.py [hunk 1]
```diff
@@ -288,6 +288,15 @@ def subs(self, *args, **kwargs):
         return self.func(*[getattr(a, 'subs', lambda x: a)(*args, **kwargs)
```
Comment (FabioLuporini) [inline devito/finite_differences/differentiable.py:288]: so this one doesn't call `_subs` directly, but, I'd guess, only via the superclass? that's a bit odd? or maybe not, actually

File: devito/finite_differences/finite_difference.py [hunk 2]
```diff
@@ -270,13 +272,10 @@ def indices_weights_to_fd(expr, dim, inds, weights, matvec=1):
     # Loop through weights
     terms = []
     for i, c in zip(inds, weights):
-        try:
-            iloc = i.xreplace(mapper)
-        except AttributeError:
-            iloc = i
-        subs = {expr.indices_ref[dim]: iloc}
-        terms.append(expr.subs(subs) * c)
+        iloc = sympify(i).xreplace(mapper)
```
Comment (FabioLuporini) [inline devito/finite_differences/finite_difference.py:275]: this block perhaps need a comment about what's going on

is the `sympify` there for the usual reasons (need for integer indices?)

Comment (FabioLuporini) [reply devito/finite_differences/finite_difference.py:275]: isn't try-catch (on top of your changes) potentially faster?

File: devito/types/basic.py [hunk 3]
```diff
@@ -770,6 +770,15 @@ def evaluate(self):
             return self
         return weight * sum(avg_list)
 
+    def _subs(self, old, new, **hints):
```
Comment (FabioLuporini) [inline devito/types/basic.py:773]: this looks identical to the one in Differentiable? Since DiscreteFunction is a subclass of both AbstractFunction and Differentiable, why don't we just drop it?

File: examples/seismic/preset_models.py [hunk 4]
```diff
@@ -219,7 +219,7 @@ def demo_model(preset, **kwargs):
                              dtype=dtype, spacing=spacing, nbl=nbl, epsilon=epsilon,
                              delta=delta, theta=theta, phi=phi, bcs="damp", **kwargs)
 
-        if kwargs.get('smooth', True):
+        if kwargs.get('smooth', False):
```
Comment (FabioLuporini) [inline examples/seismic/preset_models.py:222]: why?

do we now get smoothing by default? when I run `python tti_example.py` for example

File: devito/types/basic.py (hunk 3 above)
Comment (mloubout) [reply devito/types/basic.py:773]: This is not in Differentiable, only in DifferentiableOp.

File: examples/seismic/preset_models.py (hunk 4 above)
Comment (mloubout) [reply examples/seismic/preset_models.py:222]: Yes that's what ` kwargs.get('smooth', True)` does, currently needs to explicitly ask for no smoothing and not easy in bench/examples (need extra cli arg)

File: devito/finite_differences/finite_difference.py (hunk 2 above)
Comment (mloubout) [reply devito/finite_differences/finite_difference.py:275]: >isn't try-catch (on top of your changes) potentially faster?

Not from what I tried no. WIll add comment

File: devito/finite_differences/differentiable.py (hunk 1 above)
Comment (mloubout) [reply devito/finite_differences/differentiable.py:288]: Need subs for when there is more than one arg. `_subs` only accept `u._subs(old, new)` for dict/tuple of tuple this one is used. This is still a bit more efficient than the default sympy.expr version of it.

File: devito/types/basic.py (hunk 3 above)
Comment (mloubout) [reply devito/types/basic.py:773]: Also previously missed small cases that can happen only for a function (or symbol) where you can have `u._subs(u, v)` which is not the case for an expression.

File: devito/finite_differences/finite_difference.py (hunk 2 above)
Comment (FabioLuporini) [reply devito/finite_differences/finite_difference.py:277]: That's fine, but for what is worth, I'd use a try-catch. That's what is for, and not using it in this case only makes things potentially slower.

File: devito/types/basic.py (hunk 3 above)
Comment (FabioLuporini) [reply devito/types/basic.py:773]: I'm not sure I follow that last past... the `if old is self` at least should apply to generic exprs too?

I have a feeling, potentially wrong, that there's something off with this _subs being different than what we got in DifferentiableOp.

Other question: why is `Differentiable.subs` different than `DifferentiableOp.subs`? Should we rather try to have just one general `_subs` in `Differentiable`?

File: examples/seismic/tti/operators.py [hunk 5]
```diff
@@ -75,22 +74,18 @@ def Gzz_centered(model, field, costheta, sintheta, cosphi, sinphi, space_order):
...
-    Gzz = (first_derivative(Gz * sintheta * cosphi,
-                            dim=x, side=centered, fd_order=order1,
-                            matvec=transpose) +
-           first_derivative(Gz * sintheta * sinphi,
-                            dim=y, side=centered, fd_order=order1,
-                            matvec=transpose) +
-           first_derivative(Gz * costheta,
-                            dim=z, side=centered, fd_order=order1,
-                            matvec=transpose))
+    Gz = -(sintheta * cosphi * field.dx(fd_order=order1) +
```
Comment (FabioLuporini) [inline examples/seismic/tti/operators.py:77]: about time 😎

File: tests/test_dse.py [hunk 6]
```diff
@@ -1411,7 +1411,7 @@ def test_fullopt_w_mpi(self):
 
     @switchconfig(profiling='advanced')
     @pytest.mark.parametrize('space_order,expected', [
-        (8, 173), (16, 307)
+        (8, 173), (16, 309)
```
Comment (FabioLuporini) [inline tests/test_dse.py:1414]: how are they different now?

fun that it only happens with the higher order one

File: devito/finite_differences/finite_difference.py (hunk 2 above)
Comment (mloubout) [reply devito/finite_differences/finite_difference.py:277]: Actually it makes thig potentially better since it guaranties indices are already sympy object  for subsequent manipulation.

Comment (FabioLuporini) [reply devito/finite_differences/finite_difference.py:277]: yeah, but then you can use `sympify` only in the catch? so eventually you systematically have sympy objects

File: devito/finite_differences/differentiable.py [hunk 7]
```diff
@@ -143,6 +143,23 @@ def __getattr__(self, name):
             return self._fd[name][0](self)
         raise AttributeError("%r object has no attribute %r" % (self.__class__, name))
 
+    def _eval_subs(self, old, new, **hints):
```
Comment (FabioLuporini) [inline devito/finite_differences/differentiable.py:146]: I guess the name `_eval_subs` (instead of `_subs`) is what sympy uses underneath ? so they have `subs` (public) and `_subs` + `_eval_subs` (private) ?

Comment (mloubout) [reply devito/finite_differences/differentiable.py:146]: Forget about it, reverting to previous commit any combination of moving to differentiable is about two times slower than previously

File: devito/types/basic.py (hunk 3 above)
Comment (mloubout) [reply devito/types/basic.py:773]: > just one general _subs in Differentiable?

Way slower (no idea why) and also does weird stuff for non Add/Mul so why was originally moved to DifferentialOp

File: tests/test_dse.py (hunk 6 above)
Comment (mloubout) [reply tests/test_dse.py:1414]: Probably due to different order, the `evalf` was on the full fancy equation and may have been doing slughtly different stuff.

File: devito/types/basic.py (hunk 3 above)
Comment (FabioLuporini) [reply devito/types/basic.py:773]: so why the below 

```
        if old is self:
             return new
         if old is new:
             return self
```

doesn't apply to DifferentiableOp?

File: devito/finite_differences/differentiable.py [hunk 8]
```diff
@@ -288,6 +288,15 @@ def subs(self, *args, **kwargs):
         return self.func(*[getattr(a, 'subs', lambda x: a)(*args, **kwargs)
                            for a in self.args], evaluate=False)
 
+    def _subs(self, old, new, **hints):
```
Comment (FabioLuporini) [inline devito/finite_differences/differentiable.py:291]: out of curiosity, how is this different than the one in sympy?

File: devito/types/dense.py [hunk 9]
```diff
@@ -1327,15 +1327,15 @@ def forward(self):
         i = int(self.time_order / 2) if self.time_order >= 2 else 1
         _t = self.dimensions[self._time_position]
 
-        return self.subs({_t: _t + i * _t.spacing})
+        return self._subs(_t, _t + i * _t.spacing)
```
Comment (FabioLuporini) [inline devito/types/dense.py:1330]: out of curiosity: was this change necessary for functionality or for speed?

Comment (FabioLuporini) [reply devito/types/dense.py:1330]: ah ok, it's because now subs accepts a single pair `old,new` as two separate args

File: devito/finite_differences/differentiable.py (hunk 8 above)
Comment (FabioLuporini) [reply devito/finite_differences/differentiable.py:291]: so, I just checked sympy.Basic._subs

so now this overriding does something different, so we should probably at least update the docstring

as I wrote in another comment below, I'm still a bit baffled by the lack of a check for identity (ie something along the lines as `if self is old then ...`. 

Also check this out: https://github.com/sympy/sympy/blob/master/sympy/core/basic.py#L998

File: devito/finite_differences/differentiable.py [hunk 10]
```diff
@@ -288,6 +288,15 @@ def subs(self, *args, **kwargs):
...
+        args = list(self.args)
+        for i, arg in enumerate(args):
+            try:
```
Comment (FabioLuporini) [inline devito/finite_differences/differentiable.py:294]: OK, hang on... so is this overriding basically to avoid sympification ?

**Comment by mloubout** [comment]
>, I got a question: why do we override _subs instead of _eval_subs, which seems to be the stub we need here?

`aresame`  lot more expensive than the `old is new` in the one here.

File: devito/types/dense.py [hunk 11]
```diff
@@ -102,6 +102,8 @@ def __eq__(self, other):
         # is that they are indeed the same exact object
         return self is other
 
+    _subs = Differentiable._subs
```
Comment (FabioLuporini) [inline devito/types/dense.py:105]: this line should appear below all special methods lines ie those starting with `__` such as `__hash__`

Comment (mloubout) [reply devito/types/dense.py:105]: I prefer it here with `__eq__` that is sympy overwrite first then our stuff.

**Comment by FabioLuporini** [comment]
merged
